File: java/SwerveWithChoreo/src/main/java/frc/robot/AutoRoutines.java
package frc.robot;

import choreo.auto.AutoFactory;
import choreo.auto.AutoLoop;
import choreo.auto.AutoTrajectory;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.subsystems.CommandSwerveDrivetrain;

public class AutoRoutines {
    private final CommandSwerveDrivetrain m_drivetrain;

    public AutoRoutines(CommandSwerveDrivetrain drivetrain) {
        m_drivetrain = drivetrain;
    }

    public Command simplePathAuto(AutoFactory factory) {
        final AutoLoop routine = factory.newLoop("SimplePath Auto");
        final AutoTrajectory simplePath = factory.trajectory("SimplePath", routine);

        routine.enabled().onTrue(
            m_drivetrain.runOnce(() ->
                m_drivetrain.resetPose(
                    simplePath.getInitialPose().orElseGet(() -> {
                        routine.kill();
                        return new Pose2d();
                    })
                )
            ).andThen(simplePath.cmd())
        );
        return routine.cmd();
    }
}

File: cpp/SwerveWithChoreo/src/main/cpp/AutoRoutines.cpp
#include "AutoRoutines.h"

frc2::CommandPtr AutoRoutines::SimplePathAuto(choreo::AutoFactory<choreo::SwerveSample> &factory)
{
    auto routine = std::make_shared<choreo::AutoLoop<choreo::SwerveSample>>(
        factory.NewLoop("SimplePath Auto")
    );
    auto simplePath = std::make_shared<choreo::AutoTrajectory<choreo::SwerveSample>>(
        factory.Trajectory("SimplePath", *routine)
    );

    routine->Enabled().OnTrue(
        m_drivetrain.RunOnce([=] {
            auto const pose = simplePath->GetInitialPose();
            if (pose) {
                m_drivetrain.ResetPose(*pose);
            } else {
                routine->Kill();
            }
        })
        .AndThen(simplePath->Cmd())
    );
    return routine->Cmd();
}
